doqlistener.go

```diff
@@ -128,31 +126,31 @@ func (s DoQListener) handleConnection(connection quic.Connection) {
 	s.metrics.connection.Add(1)
 
 	for {
-		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second) // TODO: configurable
+		ctx := context.Background() // TODO: configurable
 		stream, err := connection.AcceptStream(ctx)
 		if err != nil {
-			cancel()
+			CloseConn(connection, log)
 			break
 		}
 		log.WithField("stream", stream.StreamID()).Trace("opening stream")
 		go func() {
-			s.handleStream(stream, log, ci)
-			cancel()
+			s.handleStream(stream, log, ci, connection)
 			log.WithField("stream", stream.StreamID()).Trace("closing stream")
+			_ = stream.Close()
 		}()
 	}
 }
 
-func (s DoQListener) handleStream(stream quic.Stream, log *logrus.Entry, ci ClientInfo) {
+func (s DoQListener) handleStream(stream quic.Stream, log *logrus.Entry, ci ClientInfo, connection quic.Connection) {
 	// DNS over QUIC uses one stream per query/response.
-	defer stream.Close()
 	s.metrics.stream.Add(1)
 
 	// DoQ requires a length prefix, like TCP
 	var length uint16
 	if err := binary.Read(stream, binary.BigEndian, &length); err != nil {
 		s.metrics.err.Add("read", 1)
 		log.WithError(err).Error("failed to read query")
+		CloseConn(connection, log)
```

Comment:
Not a fan of closing the connection in multiple places, including from within the stream. One effect of doing it here will be that `connection.AcceptStream(ctx)` above will return an error, which will then call `CloseConn` a second time. It should write `"closing connection"` twice in the logs too.

The prior implementation only closed it in one place. Is that not working? If a stream fails, I'd expect `AcceptStream()` to also fail, which then closes the connection.

Reply:
I agree. I did remove the changes and updated the PR